### PR TITLE
[EH] Make std::terminate() work with EH

### LINKED ISSUE #16407: Stack overflow when calling `std::terminate` + `-fwasm-exceptions`
The following code when compiled with `-fwasm-exceptions` causes a stack overflow when executed:

```c++
// std_terminate.cc

#include <exception>

int main(int argc, char **argv)
{
    std::terminate();
    return 0;
}
```

**Version of emscripten/emsdk:**
```
emcc (Emscripten gcc/clang-like replacement + linker emulating GNU ld) 3.1.6 (c102a423a06d9833a8481d6859888bebf22842b0)
clang version 15.0.0 (https://github.com/llvm/llvm-project 6eec4835844439ab932515ff4ac857773c639171)
Target: wasm32-unknown-emscripten
Thread model: posix
InstalledDir: /Users/gabriel/Dev/emsdk/upstream/bin
```

**Failing command line in full:**
```
emcc std_terminate.cc -fwasm-exceptions -o std_terminate.js -g
node --experimental-wasm-eh std_terminate.js
```

This is the end of the program's output:
```
...
Aborted(native code called abort())
Aborted(native code called abort())
Aborted(native code called abort())
Aborted(native code called abort())
Aborted(native code called abort())
Aborted(native code called abort())
Aborted(native code called abort())
/Users/gabriel/Dev/emscripten-experiments/std_terminate.js:146
      throw ex;
      ^

RangeError: Maximum call stack size exceeded
    at console.value (node:internal/console/constructor:289:13)
    at console.warn (node:internal/console/constructor:364:26)
    at abort (/Users/gabriel/Dev/emscripten-experiments/std_terminate.js:1468:3)
    at _abort (/Users/gabriel/Dev/emscripten-experiments/std_terminate.js:1828:7)
    at abort_message (wasm://wasm/000abcea:wasm-function[11]:0x48b)
    at demangling_terminate_handler() (wasm://wasm/000abcea:wasm-function[12]:0x553)
    at std::__terminate(void (*)()) (wasm://wasm/000abcea:wasm-function[17]:0x601)
    at std::terminate() (wasm://wasm/000abcea:wasm-function[15]:0x5db)
    at std::__terminate(void (*)()) (wasm://wasm/000abcea:wasm-function[17]:0x626)
    at std::terminate() (wasm://wasm/000abcea:wasm-function[15]:0x5db)
```

**Full link command and output with `-v` appended:**
```
  "/Users/gabriel/Dev/emsdk/upstream/bin/clang++" -target wasm32-unknown-emscripten -DEMSCRIPTEN -D__EMSCRIPTEN_major__=3 -D__EMSCRIPTEN_minor__=1 -D__EMSCRIPTEN_tiny__=6 -fvisibility=default -mllvm -combiner-global-alias-analysis=false -mllvm -enable-emscripten-sjlj -mllvm -disable-lsr -D_LIBCPP_ABI_VERSION=2 -fno-threadsafe-statics -Werror=implicit-function-declaration -Xclang -iwithsysroot/include/SDL --sysroot=/Users/gabriel/Dev/emsdk/upstream/emscripten/cache/sysroot -Xclang -iwithsysroot/include/compat -fwasm-exceptions -g -v std_terminate.cc -c -o /var/folders/qf/gn4cjsnx0pj84fdxgzgt780c0000gn/T/emscripten_temp_z8ny51_l/std_terminate_0.o
clang version 15.0.0 (https://github.com/llvm/llvm-project 6eec4835844439ab932515ff4ac857773c639171)
Target: wasm32-unknown-emscripten
Thread model: posix
InstalledDir: /Users/gabriel/Dev/emsdk/upstream/bin
 (in-process)
 "/Users/gabriel/Dev/emsdk/upstream/bin/clang-15" -cc1 -triple wasm32-unknown-emscripten -emit-obj -mrelax-all --mrelax-relocations -disable-free -clear-ast-before-backend -disable-llvm-verifier -discard-value-names -main-file-name std_terminate.cc -mrelocation-model static -mframe-pointer=none -ffp-contract=on -fno-rounding-math -mconstructor-aliases -target-feature +exception-handling -mllvm -wasm-enable-eh -target-cpu generic -mllvm -treat-scalable-fixed-error-as-warning -debug-info-kind=constructor -dwarf-version=4 -debugger-tuning=gdb -target-linker-version 609.8 -v -fcoverage-compilation-dir=/Users/gabriel/Dev/emscripten-experiments -resource-dir /Users/gabriel/Dev/emsdk/upstream/lib/clang/15.0.0 -D EMSCRIPTEN -D __EMSCRIPTEN_major__=3 -D __EMSCRIPTEN_minor__=1 -D __EMSCRIPTEN_tiny__=6 -D _LIBCPP_ABI_VERSION=2 -isysroot /Users/gabriel/Dev/emsdk/upstream/emscripten/cache/sysroot -internal-isystem /Users/gabriel/Dev/emsdk/upstream/emscripten/cache/sysroot/include/wasm32-emscripten/c++/v1 -internal-isystem /Users/gabriel/Dev/emsdk/upstream/emscripten/cache/sysroot/include/c++/v1 -internal-isystem /Users/gabriel/Dev/emsdk/upstream/lib/clang/15.0.0/include -internal-isystem /Users/gabriel/Dev/emsdk/upstream/emscripten/cache/sysroot/include/wasm32-emscripten -internal-isystem /Users/gabriel/Dev/emsdk/upstream/emscripten/cache/sysroot/include -Werror=implicit-function-declaration -fdeprecated-macro -fdebug-compilation-dir=/Users/gabriel/Dev/emscripten-experiments -ferror-limit 19 -fvisibility default -fgnuc-version=4.2.1 -fno-threadsafe-statics -fcxx-exceptions -fexceptions -exception-model=wasm -fcolor-diagnostics -iwithsysroot/include/SDL -iwithsysroot/include/compat -mllvm -combiner-global-alias-analysis=false -mllvm -enable-emscripten-sjlj -mllvm -disable-lsr -o /var/folders/qf/gn4cjsnx0pj84fdxgzgt780c0000gn/T/emscripten_temp_z8ny51_l/std_terminate_0.o -x c++ std_terminate.cc
clang -cc1 version 15.0.0 based upon LLVM 15.0.0git default target x86_64-apple-darwin21.3.0
ignoring nonexistent directory "/Users/gabriel/Dev/emsdk/upstream/emscripten/cache/sysroot/include/wasm32-emscripten/c++/v1"
ignoring nonexistent directory "/Users/gabriel/Dev/emsdk/upstream/emscripten/cache/sysroot/include/wasm32-emscripten"
#include "..." search starts here:
#include <...> search starts here:
 /Users/gabriel/Dev/emsdk/upstream/emscripten/cache/sysroot/include/SDL
 /Users/gabriel/Dev/emsdk/upstream/emscripten/cache/sysroot/include/compat
 /Users/gabriel/Dev/emsdk/upstream/emscripten/cache/sysroot/include/c++/v1
 /Users/gabriel/Dev/emsdk/upstream/lib/clang/15.0.0/include
 /Users/gabriel/Dev/emsdk/upstream/emscripten/cache/sysroot/include
End of search list.
 "/Users/gabriel/Dev/emsdk/upstream/bin/wasm-ld" -o std_terminate.wasm /var/folders/qf/gn4cjsnx0pj84fdxgzgt780c0000gn/T/emscripten_temp_z8ny51_l/std_terminate_0.o -L/Users/gabriel/Dev/emsdk/upstream/emscripten/cache/sysroot/lib/wasm32-emscripten -lGL -lal -lhtml5 -lstubs-debug -lnoexit -lc-debug -ldlmalloc -lcompiler_rt -lc++-except -lc++abi-except -lunwind-except -lsockets -mllvm -combiner-global-alias-analysis=false -mllvm -enable-emscripten-sjlj -mllvm -disable-lsr -mllvm -wasm-enable-eh -mllvm -exception-model=wasm --import-undefined --export-if-defined=main --export-if-defined=__start_em_asm --export-if-defined=__stop_em_asm --export-if-defined=__stdio_exit --export=emscripten_stack_get_end --export=emscripten_stack_get_free --export=emscripten_stack_get_base --export=emscripten_stack_init --export=stackSave --export=stackRestore --export=stackAlloc --export=__wasm_call_ctors --export=__errno_location --export-table -z stack-size=5242880 --initial-memory=16777216 --no-entry --max-memory=16777216 --global-base=1024
 "/Users/gabriel/Dev/emsdk/upstream/bin/wasm-emscripten-finalize" --minimize-wasm-changes -g --dyncalls-i64 --dwarf std_terminate.wasm -o std_terminate.wasm --detect-features
 "/Users/gabriel/Dev/emsdk/node/14.18.2_64bit/bin/node" /Users/gabriel/Dev/emsdk/upstream/emscripten/src/compiler.js /var/folders/qf/gn4cjsnx0pj84fdxgzgt780c0000gn/T/tmpd78xv3y0.json
 "/Users/gabriel/Dev/emsdk/upstream/bin/llvm-objcopy" std_terminate.wasm std_terminate.wasm --remove-section=producers
```

### PR DESCRIPTION
**Pull request by aheejin**
`noexcept` function shouldn't throw, so `noexcept` function code
generation is to `invoke` every function call in those functions and in
case they throw, call `std::terminate`. This codegen comes from clang
and native platforms do this too. So in wasm, they become something like
```wasm
try
  function body
catch_all
  call std::terminate
end
```

`std::terminate` calls `std::__terminate`. Both of `std::terminate` and
`std::__terminate` are `noexcept` now. So that means their code is
structured like that, which sounds like self-calling, but normally no
function calls in those functions should ever throw, so that's fine. But
in our case, `abort` ends up throwing, which is a problem.

The function body of `__terminate` eventually calls JS `abort`, and ends
up here:
https://github.com/emscripten-core/emscripten/blob/970998b2670a9bcf39d31e2b01db571089955add/src/preamble.js#L605-L623

This ends up throwing a JS exception. This is basically just a foreign
exception from the wasm perspective, and is caught by `catch_all`, and
calls `std::terminate` again. And the whole process continues until the
call stack is exhausted.

What #9730 tried to do was throwing a trap, because Wasm
`catch`/`catch_all` don't catch traps. Traps become `RuntimeError`s
after they hit a JS frame. To be consistent, we decided
`catch`/`catch_all` shouldn't catch them after they become
`RuntimeError`s. That's the reason #9730 changed the code to throw not
just any random thing but `RuntimeError`. But somehow we decided that we
make that trap distinction not based on `RuntimeError` class but some
hidden field
(https://github.com/WebAssembly/exception-handling/issues/89#issuecomment-559533953).

This PR removes `noexcept` from `std::terminate` and
`std::__terminate`'s signatures so that the cleanup that contains
`catch_all` is not generated for those two functions. So now the JS
exception thrown by `abort()` will unwind the stack, which is different
from native, but that can be considered OK because I don't think users
expect `abort` to preserve the stack intact?

Fixes #16407.

### CODE REVIEW

**Comment by aheejin** [comment]
This is an alternative approach to #16910.

File: system/lib/libcxx/include/exception [hunk 1]
```diff
@@ -45,7 +45,11 @@ unexpected_handler get_unexpected() noexcept;
 typedef void (*terminate_handler)();
 terminate_handler set_terminate(terminate_handler  f ) noexcept;
 terminate_handler get_terminate() noexcept;
+#ifdef __USING_WASM_EXCEPTIONS__
```
Comment (kripken) [inline system/lib/libcxx/include/exception:48]: Might want to add a comment here, or just point people to libcxxabi for an explanation. Unless that's obvious enough?

**Review by kripken — COMMENTED** [review]
Some questions:

1. Are there downsides to removing `noexcept`? I'd guess it prevents some LLVM optimizations, maybe mostly in LTO, but probably `[[noreturn]]` is good enough anyhow?
2. Can we get rid of the try-catch for noexcept functions in release builds? I understand from you that they appear in native builds too. But it seems like they handle an error case by converting one type of error (an exception thrown where one shouldn't be) into another (a terminate), and I'd hope that a release build would prefer to not add overhead for that. I guess in theory an exception could be thrown that is caught by the application higher up, so the try-catch converts potentially incorrect internal operations into a clear immediate terminate, but still, that feels odd to me in a release build - it feels like assertions used in debug builds.

**Comment by aheejin** [comment]
> Some questions:
> 
> 1. Are there downsides to removing `noexcept`? I'd guess it prevents some LLVM optimizations, maybe mostly in LTO, but probably `[[noreturn]]` is good enough anyhow?

I don't think it would harm optimizations. The frontend doesn't think `std::terminate` throws anyway, which I think is a baked-in assumption.

> 2. Can we get rid of the try-catch for noexcept functions in release builds? I understand from you that they appear in native builds too. But it seems like they handle an error case by converting one type of error (an exception thrown where one shouldn't be) into another (a terminate), and I'd hope that a release build would prefer to not add overhead for that. I guess in theory an exception could be thrown that is caught by the application higher up, so the try-catch converts potentially incorrect internal operations into a clear immediate terminate, but still, that feels odd to me in a release build - it feels like assertions used in debug builds.

But that will make behavior different. `std::terminate` might be OK because it is a special case, but if we do that for all `noexcept` user functions,
```cpp
void foo() noexcept {
  throw 3;
}

int main() {
  try {
    foo();
  } catch (...) {
    printf("I shouldn't be printed");
  }
}
```
"I shouldn't be printed" will not be printed in debug builds but will be printed in release builds.

---

But I became more skeptical about this approach; maybe we shouldn't run destructors after `abort` after all and land #16910.

File: system/lib/libcxx/include/exception (hunk 1 above)
Comment (aheejin) [reply system/lib/libcxx/include/exception:48]: I was not sure if it would be OK to copy-paste the same comment four times. Did that anyway.

**Comment by kripken** [comment]
> "I shouldn't be printed" will not be printed in debug builds but will be printed in release builds.

Oh, right, this is more general than terminate... Yes, my suggestion wasn't useful then.

I lean towards #16910 too, but I don't have a strong preference.

**Comment by aheejin** [comment]
Closing in favor of #16910.

**Pull request closed without merging**